bpo-47189: What's New in 3.11: Faster CPython (#32235)
python/cpython: Doc/whatsnew/3.11.rst

.. This section singles out the most important changes in Python 3.11.
   Brevity is key.

> - Python 3.11 is 22% faster than Python 3.10. See `Faster CPython`_ for details.

Reviewer: Maybe add something like "on the standard benchmark suite"? Just so people don't get upset if whatever they're doing doesn't get faster.

> - Python 3.11 is 22% faster than Python 3.10. See `Faster CPython`_ for details.

Reviewer 2: I would rather claim a range, say 10% to 60%.

Once the latest PRs are merged it will be about 25%, if you must have a single number.

https://gist.github.com/markshannon/bf1b99177290db30dd0ab8f56446d0ea

Author: I will update this number when the benchmark link updates its values/ when those PRs are merged into `main`.

.. This section singles out the most important changes in Python 3.11.
   Brevity is key.

> - Python 3.11 is 10-60% faster than Python 3.10. On average, we measured a 1.22x

Reviewer 3: ```suggestion
- Python 3.11 is up to 60% faster than Python 3.10. On average, we measured a 1.22x
```

  (Contributed by Inada Naoki in :issue:`46845`.)

Faster CPython
==============

CPython 3.11 is on average `1.22x faster <https://github.com/faster-cpython/ideas/blob/main/main-vs-310.rst>`_
than CPython 3.10 when measured with the
`pyperformance <https://github.com/python/pyperformance>`_ benchmark suite,
and compiled with GCC on Ubuntu Linux. Depending on your workload, the speedup

> could be 10-60% faster.

Reviewer 3: ```suggestion
could be up to 60% faster.
```
This phrasing seems safer, since some benchmarks actually got slower (though we're still looking into *why*).

This project focuses on two major areas in Python: faster startup and faster
runtime. Other optimizations not under this project are listed in `Optimizations`_.

Faster Startup
--------------

Frozen imports / Static code objects
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

Python caches bytecode in the :ref:`__pycache__<tut-pycache>` directory to
speed up module loading.

Previously in 3.10, Python module execution looked like this:

.. code-block:: text

   Read __pycache__ -> Unmarshal -> Heap allocated code object -> Evaluate

In Python 3.11, the core modules essential for Python startup are "frozen".
This means that their code objects (and bytecode) are statically allocated
by the interpreter. This reduces the steps in module execution process to this:

.. code-block:: text

   Statically allocated code object -> Evaluate

Interpreter startup is now 10-15% faster in Python 3.11. This has a big
impact for short-running programs using Python such as ``python -m venv ...``,

> or ``python -m pip ...``.

Reviewer 3: Those aren't great examples, since both of these tend to do a lot of I/O that will dwarf the extra startup. In fact I'm not sure I can think of great examples -- maybe we should just cut the "such as ..." part of this sentence.

(Contributed by Eric Snow, Guido van Rossum and Kumar Aditya in numerous issues.)

Faster Runtime
--------------

Cheaper, lazy Python frames
~~~~~~~~~~~~~~~~~~~~~~~~~~~
Python frames are created whenever Python calls a Python function. This frame
holds execution information. The following are new frame optimizations:

- Streamlined the frame creation process.
- Avoided memory allocation by generously re-using frame space on the C stack.

> - Streamlined the internal frame object to contain only essential information.

Reviewer 3: ```suggestion
- Streamlined the internal frame struct to contain only essential information.
```

  Frames previously held extra debugging and memory management information.

Old-style frames are now created only when required by debuggers. For most

> user code, no frames are created at all. As a result, nearly all Python

Reviewer 3: ```suggestion
Old-style frame objects are now created only when required by debuggers. For most
user code, no frame objects are created at all. As a result, nearly all Python
```

functions calls have sped up significantly. We measured a 3-7% speedup
in pyperformance.

(Contributed by Mark Shannon in :issue:`44590`.)

.. _inline-calls:

Inlined Python function calls
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
During a Python function call, Python will call a C function to interpret that
function's code, similar to how a user would call ``eval`` to run arbitrary

> code.

Reviewer 3: I'm not sure that mentioning `eval()` is very helpful here, but I do think we could mention the implication of using a C function for recursion. Here's a suggestion:

```suggestion
function's code.
This effectively limits pure Python recursion to what's safe for the C stack.
```

> In 3.11, when Python detects code calling another Python function,

Reviewer 3: ```suggestion
In 3.11, when Python detects Python code calling another Python function,
```

it sets up a new frame, and "jumps" to the new code inside the new frame. This
avoids calling the C interpreting function altogether.

Python function calls now consume almost no C stack space. This speeds up Python

> to Python function calls. In simple recursive functions like fibonacci or

Reviewer 3: ```suggestion
Python-to-Python calls now consume no C stack space.
This speeds up most such calls.
In simple recursive functions like Fibonacci or
```

Author: I'm not sure about removing the "almost". Isn't C stacked space still used for the `_InterpreterFrame` structs?

Reviewer 2: Either we make a call in C, which uses a lot of C stack space to get back to `_PyEval_EvalFrameDefault()`, or we don't and it uses none. We never use `alloca` so C stack use is all or nothing in this case.

The only Python-to-Python calls that make C calls are the "disguised" ones, e.g. calling `def __add__()` from `a + b`.

factorial, a 1.7x speedup was observed. This also means recursive functions
can recurse significantly deeper, assuming the recursion limit and memory limit

> is not exceeded. We measured a 1-3% improvement in pyperformance.

Reviewer 3: ```suggestion
can recurse significantly deeper (if the user increases the recursion limit).
We measured a 1-3% improvement in pyperformance.
```

(Contributed by Pablo Galindo and Mark Shannon in :issue:`45256`.)

PEP 659: Specializing Adaptive Interpreter
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
:pep:`659` is one of the key parts of the faster CPython project. The general
idea is that while Python is a dynamic language, most code has regions where
objects and types rarely change. This concept is known as *type stability*.

At runtime, Python will try to look for common patterns and type stability
in the executing code. Python will then replace the current operation with a
more specialized one. This specialized operation uses fast paths available only
to those use cases/types, which generally outperform their generic
counterparts. This also brings in another concept called *inline caching*, where
Python caches the results of expensive operations directly in the bytecode.

The specializer will also combine certain common instruction pairs into one
superinstruction. This reduces the overhead during execution.

This extra information requires more memory. Python will only specialize

> when it sees code that is "hot" (executed multiple times). This prevents Python

Reviewer 3: This information already seems to be out of date. :-( Maybe we should be a little more vague to avoid having to rewrite this again in a few weeks?

Reviewer 2: Hmm. I think we should be precise, but that will involve a rewrite all too soon.

from wasting time for run-once code. Python can also de-specialize when code is
too dynamic or when the use changes. Specialization is attempted periodically,
and specialization attempts are not too expensive. This allows specialization
to adapt to new circumstances.

(PEP written by Mark Shannon, with ideas inspired by Stefan Brunthaler.
See :pep:`659` for more information.)

..
   If I missed out anyone, please add them.

+---------------+--------------------+-------------------------------------------------------+-------------------+-------------------+
| Operation     | Form               | Specialization                                        | Operation speedup | Contributor(s)    |
|               |                    |                                                       | (up to)           |                   |
+===============+====================+=======================================================+===================+===================+

> | Binary        | ``o+o; o*o; o-o;`` | Binary add, multiply and subtract for common types    | 10%               | Mark Shannon,     |

Reviewer 3: I find `o` a confusing letter to hold a variable (too close to `0` or an open bullet or part of a smiley :o). I think using `x` is more conventional, and there's no strong implication that it must be a number or cannot be an object. (You could use `a` for indexed objects, e.g. `a[i]`.)

Reviewer 2: I'd prefer `a+b`. `x` is a bit close to markup as well.

|               |                    | fast paths for their underlying types.                |                   | Brandt Bucher,    |
|               |                    |                                                       |                   | Dennis Sweeney    |
+---------------+--------------------+-------------------------------------------------------+-------------------+-------------------+
| Subscript     | ``o[i]``           | Subscripting container types such as ``list``,        | 10-30%            | Irit Katriel,     |
|               |                    | ``tuple`` and ``dict`` directly index the underlying  |                   | Mark Shannon      |
|               |                    | data structures. Subscripting custom ``__getitem__``  |                   |                   |
|               |                    | is also inlined similar to :ref:`inline-calls`.       |                   |                   |
+---------------+--------------------+-------------------------------------------------------+-------------------+-------------------+

> | Store         | ``o[i] = z``       | Similar to subscripting specialization above.         | 10-90%            | Dennis Sweeney    |

Reviewer 3: Isn't it suspect that this speedup is up to 90% while getitem is up to 30%? Are comparable micro-benchmarks used?

Author: Woops this is a little misleading, the 10-30% is just for the first sentence. The second sentence doesn't have any micro benchmarks IIRC. Considering it had 1.2x faster on one of the pyperformance benchmarks, the actual numbers in micro should be significantly higher.

10-30%: https://github.com/python/cpython/pull/27043#issuecomment-874744777
`__getitem__` https://github.com/python/cpython/pull/29592

| subscript     |                    |                                                       |                   |                   |
+---------------+--------------------+-------------------------------------------------------+-------------------+-------------------+
| Calls         | ``f(arg)``         | Calls to common builtin (C) functions such as ``len`` | 20%               | Mark Shannon,     |
|               |                    | and ``isinstance`` directly call their underlying C   |                   | Ken Jin           |

> |               | ``T(arg)``         | version. This avoids going through the internal       | 170%              |                   |

Reviewer 3: Assuming `T` stands for a class, maybe use `C`? (I think of `T` as a generic type variable...)

|               |                    | calling convention.                                   |                   |                   |
|               |                    |                                                       |                   |                   |
|               |                    | Calls to certain Python functions are inlined similar |                   |                   |

> |               |                    | to :ref:`inline-calls`.                               |                   |                   |

Reviewer 3: Is this an oblique reference to `PRECALL_NO_KW_METHOD_DESCRIPTOR_O`? Or does it refer to something else?

Author: No, it just means that `CALL_FUNCTION_EX` for things like `f(1, 2, *args, **kwargs)` aren't inlined.

Reviewer 3: Hm. It's rather subtle what you intend to say then. The text linked under `inline-calls` already describes Python-to-Python calls. But they are described here again because specialization adds... what again? It feels like claiming credit for the same optimization in two places. I do realize that there are two separate code changes, but to the readers of a high-level document like this that seems a bit too subtle.

Also, it's not clear for what operation you are claiming a 170% speedup (so a factor 2.7). The table layout makes it look like `C(arg)`. But surely the speedup for class constructors is not a single extremely large number?

|               |                    |                                                       |                   |                   |
+---------------+--------------------+-------------------------------------------------------+-------------------+-------------------+

> | Load          | ``print; len``     | The object's index in the globals/builtins namespace  | - [1]_            | Mark Shannon      |

Reviewer 3: The `print; len` example makes it think it only applies to builtins. Or that the semicolon is relevant.

Sadly I'm not sure how else to quickly summarize this. Maybe `gbl`, being a random global variable?

| attribute     |                    |                                                       | in pyperformance  |                   |
+---------------+--------------------+-------------------------------------------------------+-------------------+-------------------+
| Unpack        | ``*seq``           | Specialized for common containers such as ``list``    | 8%                | Brandt Bucher     |
| Sequence      |                    | and ``tuple``. Avoids internal calling convention.    |                   |                   |
+---------------+--------------------+-------------------------------------------------------+-------------------+-------------------+

.. [1] A similar optimization already existed since Python 3.8.  3.11
       specializes for more forms and reduces some overhead.

.. [2] A similar optimization already existed since Python 3.10.
       3.11 specializes for more forms. Furthermore, all attribute loads should
       be sped up by :issue:`45947`.

.. [3] Classes with longer inheritance chains will see greater speedups.
       This optimization effectively makes method lookup constant time

>        regardless of inheritance.

Reviewer 3: I thought the type_cache (see typeobject.c) was also O(1)? Just with a larger K. :-)

Author: Oh yeah. I'm not sure why I saw speedups for longer chains then. Hmm this is confusing. I'll just remove this altogether.

Reviewer 2: Inline caches get much better hit rates than global caches. So there will be some cases where we get O(1) performance for the inline cache and O(n) for the global cache. 
I'd change "will see greater" to "may see greater".

Misc
----

* Objects now require less memory due to lazily created object namespaces. Their
  namespace dictionaries now also share keys more freely.

>   (Contributed Mark Shannon in :issue:`45340` and :issue:`40116`.)

Reviewer 4: Perhaps under misc you could mention [the exc_info business](https://github.com/faster-cpython/ideas/issues/106#issuecomment-990172363). Something like:

* A more concise representation of exceptions in the interpreter reduced the time of catching an exception by about 10%.

FAQ
---

| Q: How should I write my code to utilize these speedups?
|
| A: You don't have to change your code. Write Pythonic code that follows common
  best practices. The Faster CPython project optimizes for common code
  patterns we observe.
|
|
| Q: Will CPython 3.11 use more memory?
|

> | A: Maybe not. We don't expect memory use to exceed 20% more versus 3.10.

Reviewer 3: ```suggestion
| A: Maybe not. We don't expect memory use to exceed 20% more than 3.10.
```

Reviewer 2: Let's finish the changes to the code object and makes some measurements before we put any numbers here.

And make it clear that the memory changes come from a number of different places:
* Virtual object dictionaries: Less
* Smaller dictionaries for string-only keys: Less
* Inline caching: More
* Compact debug info (to do): Less
* More freezing and deep-freeze: (A bit) more?
* Others I've forgotten?

  This is offset by memory optimizations for frame objects and object
  dictionaries as mentioned above.
|
|
| Q: I don't see any speedups in my workload. Why?
|
| A: Certain code won't have noticeable benefits. If your code spends most of
  its time on I/O operations, or already does most of its
  computation in a C extension library like numpy, there won't be significant
  speedup. This project currently benefits pure-Python workloads the most.
|
| Furthermore, the pyperformance figures are a geometric mean. Even within the
  pyperformance benchmarks, certain benchmarks have slowed down slightly, while
  others have sped up by nearly 2x!
|
|
| Q: Is there a JIT compiler?
|
| A: No. We're still exploring other optimizations.

About
-----

Faster CPython explores optimizations for :term:`CPython`. The main team is
funded by Microsoft to work on this full-time. The team also collaborates

> extensively with volunteer contributors in the community.

Reviewer 3: Thank you for the call out!

> extensively with volunteer contributors in the community.

Reviewer 5: Let's mention my employer here as well as they are funding part-time our collaboration with the project

```suggestion
extensively with volunteer contributors in the community. Pablo Galindo Salgado is also funded by Bloomberg LP to work on the project part-time together with the main team at Microsoft and the rest of the community.
```

Author: Woops, forgot that they provided our benchmark machines too! (Also I didn't know you were paid to work on this part-time).

Reviewer 5: No problem! Thanks for the fix :)

| Subscript     | ``a[i]``           | Subscripting container types such as ``list``,        | 10-30%            | Irit Katriel,     |
|               |                    | ``tuple`` and ``dict`` directly index the underlying  |                   | Mark Shannon      |
|               |                    | data structures.                                      |                   |                   |
|               |                    |                                                       |                   |                   |
|               |                    | Subscripting custom ``__getitem__``                   |                   |                   |
|               |                    | is also inlined similar to :ref:`inline-calls`.       |                   |                   |
+---------------+--------------------+-------------------------------------------------------+-------------------+-------------------+
| Store         | ``a[i] = z``       | Similar to subscripting specialization above.         | 10-90%            | Dennis Sweeney    |

> | subscript     |                    |                                                       |                   |                   |

Reviewer 3: So IIUC from your comment (so many comments! :-( ) the problem isn't that 90% seems high, it is that 30% is low?! I worry that we're claiming extreme improvements here that nobody will be able to repro in real life. Irit's benchmark for `__getitem__` seems to only indicate up to 25%.

And using a timeit microbenchmark similar to Irit's I can't show more than a modest speed improvement for subscript assignment:
```
PS C:\Users\gvanrossum\cpython> py -3.10 -m timeit -s "a = [0,1,2,3,4,5]" "a[0] = 0"
10000000 loops, best of 5: 22.9 nsec per loop

PS C:\Users\gvanrossum\cpython> .\installed\python.exe -m timeit -s "a = [0,1,2,3,4,5]" "a[0] = 0"
20000000 loops, best of 5: 18.9 nsec per loop
```

Reviewer 6: Looking back, I think my benchmarks were like this, so an even finer level of micro-ness. I guess this is the tragedy of microbenchmarks.

```python
from pyperf import Runner
runner = Runner()

runner.timeit("dict[str]=...",
    setup="from itertools import repeat; d = dict()",
    stmt="for s in repeat('foobar', 100_000):\n"
         "    d[s] = d[s] = d[s] = d[s] = d[s] = d[s] = d[s] = d[s] = d[s] = d[s] = s"
)
runner.timeit("list[int]=...",
    setup="from itertools import repeat; L = [0]*10",
    stmt="for i in repeat(2, 100_000):\n"
         "    L[i] = L[i] = L[i] = L[i] = L[i] = L[i] = L[i] = L[i] = L[i] = L[i] = i"
)
```

Author: Yeah I think I based it off of Dennis' microbenchmarks which aren't compatible  fully with Irit's. I will reduce it to a more moderate number (25%).

|               |                    | and ``isinstance`` directly call their underlying C   |                   | Ken Jin           |
|               | ``C(arg)``         | version. This avoids going through the internal       | 170%              |                   |
|               |                    | calling convention.                                   |                   |                   |
|               |                    |                                                       |                   |                   |
|               |                    | Calls to certain Python functions are inlined similar |                   |                   |
|               |                    | to :ref:`inline-calls`.                               |                   |                   |
|               |                    |                                                       |                   |                   |
+---------------+--------------------+-------------------------------------------------------+-------------------+-------------------+

> | Load          | ``print``          | The object's index in the globals/builtins namespace  | - [1]_            | Mark Shannon      |

Reviewer 3: ```suggestion
| Load          | ``print``          | The object's index in the globals/builtins namespace  | [1]_              | Mark Shannon      |
```

| global        | ``len``            | is cached. Loading globals and builtins require       |                   |                   |
| variable      |                    | zero namespace lookups.                               |                   |                   |
+---------------+--------------------+-------------------------------------------------------+-------------------+-------------------+

> | Load          | ``o.attr``         | Similar to loading global variables. The attribute's  | - [2]_            | Mark Shannon      |

Reviewer 3: ```suggestion
| Load          | ``o.attr``         | Similar to loading global variables. The attribute's  | [2]_              | Mark Shannon      |
```

  (Contributed by Inada Naoki in :issue:`46845`.)

Faster CPython
==============

> CPython 3.11 is on average `1.22x faster <https://github.com/faster-cpython/ideas/blob/main/main-vs-310.rst>`_

Reviewer 2: See above comment

This resulted in a 3-7% improvement in pyperformance.
(Contributed by Mark Shannon in :issue:`44590`.)

.. _inline-calls:

Inlined Python function calls
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
During a Python function call, Python will call a C function to interpret that
function's code, similar to how a user would call ``eval`` to run arbitrary
code.

In 3.11, when Python detects code calling another Python function,
it sets up a new frame, and "jumps" to the new code inside the new frame. This
avoids calling the C interpreting function altogether.

Python function calls now consume almost no C stack space. This speeds up Python
to Python function calls. In simple recursive functions like fibonacci or
factorial, a 1.7x speedup was observed. This also means recursive functions
can recurse significantly deeper, assuming the recursion limit and memory limit
is not exceeded. This resulted in a 1-3% improvement in pyperformance.
(Contributed by Pablo Galindo and Mark Shannon in :issue:`45256`.)

PEP 659: Specializing Adaptive Interpreter
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
:pep:`659` is one of the key parts of the faster CPython project. The general

> idea is that while Python is a dynamic language, most code have regions where

Reviewer: ```suggestion
idea is that while Python is a dynamic language, most code has regions where
```

objects and types rarely change. This concept is known as *type stability*.

At runtime, Python will try to look for common patterns and type stability
in the executing code. Python will then replace the current operation with a

> more specialized one. This specialized operation use fast paths available only
> to those use cases/types and will generally outperform their generic

Reviewer: ```suggestion
more specialized one. This specialized operation uses fast paths available only
to those use cases/types, which will generally outperform their generic
```
Also maybe remove "generally"

Author: Thanks for fixing those embarrassing grammar mistakes.

I say "generally" because it's true :). There are some cases where specialized opcodes don't speed up or are slower (e.g. generic vectorcalled C functions, or polymorphic code).

counterparts. This also brings in another concept called *inline caching*, where
Python caches the results of expensive operations directly in the bytecode.

The specializer will also combine certain common instruction pairs into one

> super instruction. This reduces the overhead during execution.

Reviewer: I've usually seen "superinstruction" written as one word

+---------------+--------------------+-------------------------------------------------------+-------------------+-------------------+

.. [1] Note that a similar optimization already existed since Python 3.8.

.. [2] Note that a similar optimization has already existed since Python 3.10.

.. [3] Classes with longer inheritance chains will see greater speedups.
       This optimization effectively makes method lookup constant time
       regardless of inheritance.

FAQ
---

| Q: How should I write my code to utilize these speedups?
|
| A: You don't have to change your code. Write Pythonic code that follows common
  best practices. The Faster CPython project optimizes for common code
  patterns we observe.
|
|
| Q: Will CPython 3.11 use more memory?
|
| A: Yes. However, how much exactly depends on how much code is "hot". We don't
  expect memory use to exceed 20% more versus 3.10. This may be offset by
  memory optimizations for frame objects and object dictionaries as mentioned
  above.
|
|
| Q: I don't see any speedups in my workload. Why?
|
| A: Certain code won't have noticeable benefits. If your code spends most of

>   its time on IO(Input/Output) operations, or already does most of its

Reviewer: ```suggestion
  its time on IO operations, or already does most of its
```
Or maybe write out "input/output". I feel like "IO" is common enough that it's more recognizable as an abbreviation (just like JIIT below), but don't feel strongly.

Reviewer 2: `I/O` maybe

can recurse significantly deeper, assuming the recursion limit and memory limit
is not exceeded. This resulted in a 1-3% improvement in pyperformance.
(Contributed by Pablo Galindo and Mark Shannon in :issue:`45256`.)

PEP 659: Specializing Adaptive Interpreter
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
:pep:`659` is one of the key parts of the faster CPython project. The general
idea is that while Python is a dynamic language, most code have regions where
objects and types rarely change. This concept is known as *type stability*.

At runtime, Python will try to look for common patterns and type stability
in the executing code. Python will then replace the current operation with a
more specialized one. This specialized operation use fast paths available only
to those use cases/types and will generally outperform their generic
counterparts. This also brings in another concept called *inline caching*, where
Python caches the results of expensive operations directly in the bytecode.

The specializer will also combine certain common instruction pairs into one
super instruction. This reduces the overhead during execution.

> Since this exta information requires more memory, Python will only specialize

Reviewer 2: This is not true any more. We specialize in place, so we use the extra memory regardless.
We delay specialization so that we don't waste time specializing code that is only run once, like module or class code.

when it sees code that is "hot" (executed multiple times). It can also

> de-specialize when code is too dynamic. This is attempted every so often, and

Reviewer 2: Not only when too dynamic, but when the use changes. The idea is that the code adapts to the new circumstances.

|               |                    | and ``isinstance`` directly call their underlying C   |                   | Ken Jin           |
|               | ``T(arg)``         | version. This avoids going through the internal       | 170%              |                   |
|               |                    | calling convention.                                   |                   |                   |
|               |                    |                                                       |                   |                   |
|               |                    | Calls to certain Python functions are inlined similar |                   |                   |
|               |                    | to :ref:`inline-calls`.                               |                   |                   |
|               |                    |                                                       |                   |                   |
+---------------+--------------------+-------------------------------------------------------+-------------------+-------------------+

> | Load          | ``print; len``     | The object's index in the globals/builtins namespace  | 0% [1]_           | Mark Shannon      |

Reviewer 2: We get a speed up here, because we have two different specializations, and because of the dict key version which reduces the overhead. I couldn't give you a number, though

| global        |                    | is cached. Loading globals and builtins require       |                   |                   |
| variable      |                    | zero namespace lookups.                               |                   |                   |
+---------------+--------------------+-------------------------------------------------------+-------------------+-------------------+

> | Load          | ``o.attr``         | Similar to loading global variables. The attribute's  | 0% [2]_           | Mark Shannon      |

Reviewer 2: We get a bigger speed up here, largely thanks to the virtual object dictionaries. More specializations also help here.
Don't have numbers though.

| Sequence      |                    | and ``tuple``. Avoids internal calling convention.    |                   |                   |
+---------------+--------------------+-------------------------------------------------------+-------------------+-------------------+

.. [1] Note that a similar optimization already existed since Python 3.8.

.. [2] Note that a similar optimization has already existed since Python 3.10.

.. [3] Classes with longer inheritance chains will see greater speedups.
       This optimization effectively makes method lookup constant time
       regardless of inheritance.

Misc
----

* Objects now require less memory due to lazily created object namespaces. Their
  namespace dictionaries now also share keys more freely.
  (Contributed Mark Shannon in :issue:`45340` and :issue:`40116`.)

FAQ
---

| Q: How should I write my code to utilize these speedups?
|
| A: You don't have to change your code. Write Pythonic code that follows common
  best practices. The Faster CPython project optimizes for common code
  patterns we observe.
|
|
| Q: Will CPython 3.11 use more memory?
|

> | A: Yes. However, how much exactly depends on how much code is "hot". We don't

Reviewer 2: Or no.  We use a bit more memory for code objects, but less for most other objects.
Hopefully we can get a bit more data for this before the release.

  expect memory use to exceed 20% more versus 3.10. This may be offset by
  memory optimizations for frame objects and object dictionaries as mentioned
  above.
|
|
| Q: I don't see any speedups in my workload. Why?
|
| A: Certain code won't have noticeable benefits. If your code spends most of
  its time on IO(Input/Output) operations, or already does most of its
  computation in a C extension library like numpy, there won't be significant
  speedup. This project currently benefits pure-Python workloads the most.
|
| Furthermore, the pyperformance figures are a geometric mean. Even within the
  pyperformance benchmarks, certain benchmarks have slowed down slightly, while

>   others have sped up by nearly 1.9x!

Reviewer 2: `x2` is closer and it sounds better too!

|
|
| Q: Is there a JIT compiler?
|
| A: No. We're still exploring other optimizations.

> About

Reviewer 2: I'm not sure that we need a credits list. It's not a movie :slightly_smiling_face:

Author: Hah! Good point. I wrote that before adding individual `(contributed by)`s. It's now redundant.

  (Contributed by Inada Naoki in :issue:`46845`.)

Faster CPython
==============

CPython 3.11 is on average `1.22x faster <https://github.com/faster-cpython/ideas/blob/main/main-vs-310.rst>`_
than CPython 3.10 when measured with the
`pyperformance <https://github.com/python/pyperformance>`_ benchmark suite, on

> Ubuntu Linux. Depending on your workload, the speedup could be 10-60% faster.

Reviewer 7: Should we mention compiler too? https://github.com/faster-cpython/ideas/discussions/331

The project focuses on two major areas in Python: faster startup and faster
runtime. Other optimizations not under this project are listed in `Optimizations`_.

> 

Reviewer 4: Is it necessary to separate the Faster CPython section from the optimisations section? We could have it all under optimisations, with maybe a footnote thanking Microsoft for its sponsorship of some of the work.

Author: Good point - we definitely could. I only have cosmetic reasons for not doing that. Nested sections within sections really don't render well in our current scheme. See the [3.10 what's new for example](https://docs.python.org/3/whatsnew/3.10.html) .

![image](https://user-images.githubusercontent.com/28750310/161390840-1759640a-0644-48eb-9a01-5b489f5d30e1.png)

Currently, it looks like this.

![image](https://user-images.githubusercontent.com/28750310/161390878-fcd34de7-0133-4c9e-8873-77ff44c0c5a6.png)

_Logically_ it seems better to do what you said though. All this just to avoid a single level of nesting seems over the top .

Author: Wait sorry I think I misunderstood what you intended. If you meant that we could list them as bullet points under Optimizations, I'd prefer not to.  Some parts need more elaboration, and the bullet points don't help with logical separation as much as sections do.

Reviewer 4: That brings me to my second question - is whatsnew the place to elaborate so much?  People read what's new to assess the impact of upgrading to this version.  The headline perf improvement numbers belong there, but I'm not sure about all the details. 

If we create a blog-type doc for Faster CPython (and link to it from whatsnew), then in future years we can add sections to it and it becomes a self-contained history of the project.

Author: Hmm, I think this definitely could be shorter. Where do you think I can cut some of cruft? Or maybe I can move this entire section off of whats new like you mentioned.

Frankly, I'm not sure of the policy regarding what's new. My current understanding is that whatever the release manager adds is what goes. For example [Python 3.8's whatsnew](https://docs.python.org/3/whatsnew/3.8.html) went into great deal of detail for things we might not consider today as _that_ newsworthy.

Reviewer 4: The release manager is definitely the final editor of what's new.

Faster CPython
==============

CPython 3.11 is on average `1.22x faster <https://github.com/faster-cpython/ideas/blob/main/main-vs-310.rst>`_
than CPython 3.10 when measured with the
`pyperformance <https://github.com/python/pyperformance>`_ benchmark suite,
and compiled with GCC on Ubuntu Linux. Depending on your workload, the speedup
could be 10-60% faster.

This project focuses on two major areas in Python: faster startup and faster
runtime. Other optimizations not under this project are listed in `Optimizations`_.

Faster Startup
--------------

Frozen imports / Static code objects
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

Python caches bytecode in the :ref:`__pycache__<tut-pycache>` directory to
speed up module loading.

Previously in 3.10, Python module execution looked like this:

.. code-block:: text

   Read __pycache__ -> Unmarshal -> Heap allocated code object -> Evaluate

In Python 3.11, the core modules essential for Python startup are "frozen".
This means that their code objects (and bytecode) are statically allocated
by the interpreter. This reduces the steps in module execution process to this:

.. code-block:: text

   Statically allocated code object -> Evaluate

> Interpreter startup is now 10-15% faster in Python 3.11. This positively
> impacts short-running programs using Python. Such as ``python -m venv ...``, or
> ``python -m pip ...``.

Reviewer 8: ```suggestion
Interpreter startup is now 10-15% faster in Python 3.11. This has a big
impact for short-running programs using Python such as ``python -m venv ...`` or
``python -m pip ...``.
```

| methods for   |                    | loading now has no namespace lookups -- even for      |                   | Mark Shannon      |
| call          |                    | classes with long inheritance chains.                 |                   |                   |
+---------------+--------------------+-------------------------------------------------------+-------------------+-------------------+
| Store         | ``o.attr = z``     | Similar to load attribute optimization.               | 2%                | Mark Shannon      |
| attribute     |                    |                                                       | in pyperformance  |                   |
+---------------+--------------------+-------------------------------------------------------+-------------------+-------------------+
| Unpack        | ``*seq``           | Specialized for common containers such as ``list``    | 8%                | Brandt Bucher     |
| Sequence      |                    | and ``tuple``. Avoids internal calling convention.    |                   |                   |

> +---------------+--------------------+-------------------------------------------------------+-------------------+-------------------+

Reviewer 6: There's also COMPARE_OP, if we want to include that.

Author: Yep. I left those out since they're less exciting/common operations.

Inlined Python function calls
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
During a Python function call, Python will call an evaluating C function to
interpret that function's code. This effectively limits pure Python recursion to

> what's safe for the C stack

Reviewer 3: ```suggestion
what's safe for the C stack.
```

> In 3.11, when Python detects Python code calling another Python function,

Reviewer 3: ```suggestion
In 3.11, when CPython detects Python code calling another Python function,
```

it sets up a new frame, and "jumps" to the new code inside the new frame. This
avoids calling the C interpreting function altogether.

Most Python function calls now consume no C stack space. This speeds up
most of such calls. In simple recursive functions like fibonacci or
factorial, a 1.7x speedup was observed. This also means recursive functions

> can recurse significantly deeper, (if the user increases the recursion limit).

Reviewer 3: ```suggestion
can recurse significantly deeper (if the user increases the recursion limit).
```
